django/db/models/sql/compiler.py

```diff
@@ -651,7 +651,9 @@ def find_ordering_name(self, name, opts, alias=None, default_order='ASC',
                                                        order, already_seen))
             return results
         targets, alias, _ = self.query.trim_joins(targets, joins, path)
-        return [(OrderBy(t.get_col(alias), descending=descending), False) for t in targets]
```

Comment:
I've played some more here and I think the code ends up too messy this way. I'm now exploring returning a function as part of setup_joins that returns an object the compiler can handle. For fields this is equivalent to get_col. This should get around the problems around joins I was worried about. I'll push a change once I'm home from pycon.